File: src/org/kde/kdeconnect/NetworkPacket.java
/*
 * Copyright 2014 Albert Vaca Cintora <albertvaka@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License or (at your option) version 3 or any later version
 * accepted by the membership of KDE e.V. (or its successor approved
 * by the membership of KDE e.V.), which shall act as a proxy
 * defined in Section 14 of version 3 of the license.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package org.kde.kdeconnect;

import android.content.Context;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.kde.kdeconnect.Helpers.DeviceHelper;
import org.kde.kdeconnect.Plugins.PluginFactory;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.Socket;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class NetworkPacket {

    public final static int ProtocolVersion = 7;

    public final static String PACKET_TYPE_IDENTITY = "kdeconnect.identity";
    public final static String PACKET_TYPE_PAIR = "kdeconnect.pair";

    public final static int PACKET_REPLACEID_MOUSEMOVE = 0;
    public final static int PACKET_REPLACEID_PRESENTERPOINTER = 1;

    public static Set<String> protocolPacketTypes = new HashSet<String>() {{
        add(PACKET_TYPE_IDENTITY);
        add(PACKET_TYPE_PAIR);
    }};

    private long mId;
    String mType;
    private JSONObject mBody;
    private Payload mPayload;
    private JSONObject mPayloadTransferInfo;
    private volatile boolean canceled;

    private NetworkPacket() {

    }

    public NetworkPacket(String type) {
        mId = System.currentTimeMillis();
        mType = type;
        mBody = new JSONObject();
        mPayload = null;
        mPayloadTransferInfo = new JSONObject();
    }

    public boolean isCanceled() { return canceled; }
    public void cancel() { canceled = true; }

    public String getType() {
        return mType;
    }

    public long getId() {
        return mId;
    }

    //Most commons getters and setters defined for convenience
    public String getString(String key) {
        return mBody.optString(key, "");
    }

    public String getString(String key, String defaultValue) {
        return mBody.optString(key, defaultValue);
    }

    public void set(String key, String value) {
        if (value == null) return;
        try {
            mBody.put(key, value);
        } catch (Exception ignored) {
        }
    }

    public int getInt(String key) {
        return mBody.optInt(key, -1);
    }

    public int getInt(String key, int defaultValue) {
        return mBody.optInt(key, defaultValue);
    }

    public void set(String key, int value) {
        try {
            mBody.put(key, value);
        } catch (Exception ignored) {
        }
    }

    public long getLong(String key) {
        return mBody.optLong(key, -1);
    }

    public long getLong(String key, long defaultValue) {
        return mBody.optLong(key, defaultValue);
    }

    public void set(String key, long value) {
        try {
            mBody.put(key, value);
        } catch (Exception ignored) {
        }
    }

    public boolean getBoolean(String key) {
        return mBody.optBoolean(key, false);
    }

    public boolean getBoolean(String key, boolean defaultValue) {
        return mBody.optBoolean(key, defaultValue);
    }

    public void set(String key, boolean value) {
        try {
            mBody.put(key, value);
        } catch (Exception ignored) {
        }
    }

    public double getDouble(String key) {
        return mBody.optDouble(key, Double.NaN);
    }

    public double getDouble(String key, double defaultValue) {
        return mBody.optDouble(key, defaultValue);
    }

    public void set(String key, double value) {
        try {
            mBody.put(key, value);
        } catch (Exception ignored) {
        }
    }

    public JSONArray getJSONArray(String key) {
        return mBody.optJSONArray(key);
    }

    public void set(String key, JSONArray value) {
        try {
            mBody.put(key, value);
        } catch (Exception ignored) {
        }
    }

    public JSONObject getJSONObject(String key) {
        return mBody.optJSONObject(key);
    }

    public void set(String key, JSONObject value) {
        try {
            mBody.put(key, value);
        } catch (JSONException ignored) {
        }
    }

    private Set<String> getStringSet(String key) {
        JSONArray jsonArray = mBody.optJSONArray(key);
        if (jsonArray == null) return null;
        Set<String> list = new HashSet<>();
        int length = jsonArray.length();
        for (int i = 0; i < length; i++) {
            try {
                String str = jsonArray.getString(i);
                list.add(str);
            } catch (Exception ignored) {
            }
        }
        return list;
    }

    public Set<String> getStringSet(String key, Set<String> defaultValue) {
        if (mBody.has(key)) return getStringSet(key);
        else return defaultValue;
    }

    public void set(String key, Set<String> value) {
        try {
            JSONArray jsonArray = new JSONArray();
            for (String str : value) {
                jsonArray.put(str);
            }
            mBody.put(key, jsonArray);
        } catch (Exception ignored) {
        }
    }

    public List<String> getStringList(String key) {
        JSONArray jsonArray = mBody.optJSONArray(key);
        if (jsonArray == null) return null;
        List<String> list = new ArrayList<>();
        int length = jsonArray.length();
        for (int i = 0; i < length; i++) {
            try {
                String str = jsonArray.getString(i);
                list.add(str);
            } catch (Exception ignored) {
            }
        }
        return list;
    }

    public List<String> getStringList(String key, List<String> defaultValue) {
        if (mBody.has(key)) return getStringList(key);
        else return defaultValue;
    }

    public void set(String key, List<String> value) {
        try {
            JSONArray jsonArray = new JSONArray();
            for (String str : value) {
                jsonArray.put(str);
            }
            mBody.put(key, jsonArray);
        } catch (Exception ignored) {
        }
    }

    public boolean has(String key) {
        return mBody.has(key);
    }

    public String serialize() throws JSONException {
        JSONObject jo = new JSONObject();
        jo.put("id", mId);
        jo.put("type", mType);
        jo.put("body", mBody);
        if (hasPayload()) {
            jo.put("payloadSize", mPayload.payloadSize);
            jo.put("payloadTransferInfo", mPayloadTransferInfo);
        }
        //QJSon does not escape slashes, but Java JSONObject does. Converting to QJson format.
        return jo.toString().replace("\\/", "/") + "\n";
    }

    static public NetworkPacket unserialize(String s) throws JSONException {

        NetworkPacket np = new NetworkPacket();
        JSONObject jo = new JSONObject(s);
        np.mId = jo.getLong("id");
        np.mType = jo.getString("type");
        np.mBody = jo.getJSONObject("body");
        if (jo.has("payloadSize")) {
            np.mPayloadTransferInfo = jo.getJSONObject("payloadTransferInfo");
            np.mPayload = new Payload(jo.getLong("payloadSize"));
        } else {
            np.mPayloadTransferInfo = new JSONObject();
            np.mPayload = new Payload(0);
        }
        return np;
    }

    static public NetworkPacket createIdentityPacket(Context context) {

        NetworkPacket np = new NetworkPacket(NetworkPacket.PACKET_TYPE_IDENTITY);

        String deviceId = DeviceHelper.getDeviceId(context);
        try {
            np.mBody.put("deviceId", deviceId);
            np.mBody.put("deviceName", DeviceHelper.getDeviceName(context));
            np.mBody.put("protocolVersion", NetworkPacket.ProtocolVersion);
            np.mBody.put("deviceType", DeviceHelper.getDeviceType(context).toString());
            np.mBody.put("incomingCapabilities", new JSONArray(PluginFactory.getIncomingCapabilities()));
            np.mBody.put("outgoingCapabilities", new JSONArray(PluginFactory.getOutgoingCapabilities()));
        } catch (Exception e) {
            Log.e("NetworkPackage", "Exception on createIdentityPacket", e);
        }

        return np;

    }

    public void setPayload(Payload payload) { mPayload = payload; }

    public Payload getPayload() {
        return mPayload;
    }

    public long getPayloadSize() {
        return mPayload == null ? 0 : mPayload.payloadSize;
    }

    public boolean hasPayload() {
        return (mPayload != null && mPayload.payloadSize != 0);
    }

    public boolean hasPayloadTransferInfo() {
        return (mPayloadTransferInfo.length() > 0);
    }

    public JSONObject getPayloadTransferInfo() {
        return mPayloadTransferInfo;
    }

    public void setPayloadTransferInfo(JSONObject payloadTransferInfo) {
        mPayloadTransferInfo = payloadTransferInfo;
    }

    public static class Payload {
        private InputStream inputStream;
        private Socket inputSocket;
        private long payloadSize;

        public Payload(long payloadSize) {
            this((InputStream)null, payloadSize);
        }

        public Payload(byte[] data) {
            this(new ByteArrayInputStream(data), data.length);
        }

        /**
         * <b>NOTE: Do not use this to set an SSLSockets InputStream as the payload, use Payload(Socket, long) instead because of this <a href="https://issuetracker.google.com/issues/37018094">bug</a></b>
         */
        public Payload(InputStream inputStream, long payloadSize) {
            this.inputSocket = null;
            this.inputStream = inputStream;
            this.payloadSize = payloadSize;
        }

        public Payload(Socket inputSocket, long payloadSize) throws IOException {
            this.inputSocket = inputSocket;
            this.inputStream = inputSocket.getInputStream();
            this.payloadSize = payloadSize;
        }

        /**
         * <b>NOTE: Do not close the InputStream directly call Payload.close() instead, this is because of this <a href="https://issuetracker.google.com/issues/37018094">bug</a></b>
         */
        public InputStream getInputStream() { return inputStream; }
        long getPayloadSize() { return payloadSize; }

        public void close() {
            //TODO: If socket only close socket if that also closes the streams that is
            try {
                if (inputStream != null) {
                    inputStream.close();
                }
            } catch(IOException ignored) {}

            try {
                if (inputSocket != null) {
                    inputSocket.close();
                }
            } catch (IOException ignored) {}
        }
    }
}

File: src/org/kde/kdeconnect/NetworkPacket.kt
/*
 * SPDX-FileCopyrightText: 2014 Albert Vaca Cintora <albertvaka@gmail.com>
 *
 * SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only OR LicenseRef-KDE-Accepted-GPL
*/
package org.kde.kdeconnect

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.lang.RuntimeException
import java.net.Socket
import kotlin.concurrent.Volatile

class NetworkPacket private constructor(
    val id: Long,
    val type: String,
    private val mBody: JSONObject,
    var payload: Payload?,
    var payloadTransferInfo: JSONObject,
) {
    constructor(type: String) : this(
        id = System.currentTimeMillis(),
        type = type,
        mBody = JSONObject(),
        payload = null,
        payloadTransferInfo = JSONObject()
    )

    @Volatile
    var isCanceled: Boolean = false
        private set

    fun cancel() {
        isCanceled = true
    }

    // Most commons getters and setters defined for convenience
    fun getString(key: String): String {
        return mBody.optString(key, "")
    }

    fun getStringOrNull(key: String): String? {
        return if (mBody.has(key)) mBody.getString(key)
        else null
    }

    fun getString(key: String, defaultValue: String): String {
        return mBody.optString(key, defaultValue)
    }

    operator fun set(key: String, value: String?) {
        if (value == null) return
        try {
            mBody.put(key, value)
        } catch (ignored: Exception) {
        }
    }

    fun getInt(key: String): Int {
        return mBody.optInt(key, -1)
    }

    fun getInt(key: String, defaultValue: Int): Int {
        return mBody.optInt(key, defaultValue)
    }

    operator fun set(key: String, value: Int) {
        try {
            mBody.put(key, value)
        } catch (ignored: Exception) {
        }
    }

    fun getLong(key: String): Long {
        return mBody.optLong(key, -1)
    }

    fun getLong(key: String, defaultValue: Long): Long {
        return mBody.optLong(key, defaultValue)
    }

    operator fun set(key: String, value: Long) {
        try {
            mBody.put(key, value)
        } catch (ignored: Exception) {
        }
    }

    fun getBoolean(key: String): Boolean {
        return mBody.optBoolean(key, false)
    }

    fun getBoolean(key: String, defaultValue: Boolean): Boolean {
        return mBody.optBoolean(key, defaultValue)
    }

    operator fun set(key: String, value: Boolean) {
        try {
            mBody.put(key, value)
        } catch (ignored: Exception) {
        }
    }

    fun getDouble(key: String): Double {
        return mBody.optDouble(key, Double.NaN)
    }

    fun getDouble(key: String, defaultValue: Double): Double {
        return mBody.optDouble(key, defaultValue)
    }

    operator fun set(key: String, value: Double) {
        try {
            mBody.put(key, value)
        } catch (ignored: Exception) {
        }
    }

    fun getJSONArray(key: String): JSONArray? {
        return mBody.optJSONArray(key)
    }

    operator fun set(key: String, value: JSONArray?) {
        try {
            mBody.put(key, value)
        } catch (ignored: Exception) {
        }
    }

    fun getJSONObject(key: String): JSONObject? {
        return mBody.optJSONObject(key)
    }

    operator fun set(key: String, value: JSONObject?) {
        try {
            mBody.put(key, value)
        } catch (ignored: JSONException) {
        }
    }

    fun getStringSet(key: String): Set<String>? {
        val jsonArray = mBody.optJSONArray(key) ?: return null
        val list: MutableSet<String> = HashSet()
        val length = jsonArray.length()
        for (i in 0 until length) {
            try {
                val str = jsonArray.getString(i)
                list.add(str)
            } catch (ignored: Exception) {
            }
        }
        return list
    }

    fun getStringSet(key: String, defaultValue: Set<String>?): Set<String>? {
        return if (mBody.has(key)) getStringSet(key)
        else defaultValue
    }

    operator fun set(key: String, value: Set<String>) {
        try {
            val jsonArray = JSONArray()
            for (str in value) {
                jsonArray.put(str)
            }
            mBody.put(key, jsonArray)
        } catch (ignored: Exception) {
        }
    }

    fun getStringList(key: String): List<String>? {
        val jsonArray = mBody.optJSONArray(key) ?: return null
        val list: MutableList<String> = ArrayList()
        val length = jsonArray.length()
        for (i in 0 until length) {
            try {
                val str = jsonArray.getString(i)
                list.add(str)
            } catch (ignored: Exception) {
            }
        }
        return list
    }

    fun getStringList(key: String, defaultValue: List<String>?): List<String>? {
        return if (mBody.has(key)) getStringList(key)
        else defaultValue
    }

    operator fun set(key: String, value: List<String>) {
        try {
            val jsonArray = JSONArray()
            for (str in value) {
                jsonArray.put(str)
            }
            mBody.put(key, jsonArray)
        } catch (ignored: Exception) {
        }
    }

    fun has(key: String): Boolean {
        return mBody.has(key)
    }

    @Throws(JSONException::class)
    fun serialize(): String {
        val jo = JSONObject()
        jo.put("id", id)
        jo.put("type", type)
        jo.put("body", mBody)
        if (hasPayload()) {
            jo.put("payloadSize", payload!!.payloadSize)
            jo.put("payloadTransferInfo", payloadTransferInfo)
        }
        // QJSon does not escape slashes, but Java JSONObject does. Converting to QJson format.
        try {
            return jo.toString().replace("\\/", "/") + "\n"
        } catch (e : OutOfMemoryError) {
            throw RuntimeException("OOM serializing packet of type $type", e)
        }
    }

    val payloadSize: Long
        get() = payload?.payloadSize ?: 0

    fun hasPayload(): Boolean {
        val payload = payload
        return payload != null && payload.payloadSize != 0L
    }

    fun hasPayloadTransferInfo(): Boolean {
        return payloadTransferInfo.length() > 0
    }

    class Payload {
        /**
         * **NOTE: Do not close the InputStream directly call Payload.close() instead, this is because of this [bug](https://issuetracker.google.com/issues/37018094)**
         */
        val inputStream: InputStream?
        private val inputSocket: Socket?
        val payloadSize: Long

        constructor(payloadSize: Long) : this(null, payloadSize)

        constructor(data: ByteArray) : this(ByteArrayInputStream(data), data.size.toLong())

        /**
         * **NOTE: Do not use this to set an SSLSockets InputStream as the payload, use Payload(Socket, long) instead because of this [bug](https://issuetracker.google.com/issues/37018094)**
         */
        constructor(inputStream: InputStream?, payloadSize: Long) {
            this.inputSocket = null
            this.inputStream = inputStream
            this.payloadSize = payloadSize
        }

        constructor(inputSocket: Socket, payloadSize: Long) {
            this.inputSocket = inputSocket
            this.inputStream = inputSocket.getInputStream()
            this.payloadSize = payloadSize
        }

        fun close() {
            // TODO: If socket only close socket if that also closes the streams that is
            try {
                inputStream?.close()
            } catch (ignored: IOException) {
            }

            try {
                inputSocket?.close()
            } catch (ignored: IOException) {
            }
        }
    }

    companion object {
        const val PACKET_TYPE_IDENTITY: String = "kdeconnect.identity"
        const val PACKET_TYPE_PAIR: String = "kdeconnect.pair"

        @JvmStatic
        @Throws(JSONException::class)
        fun unserialize(s: String): NetworkPacket {
            val jo = JSONObject(s)
            val id = jo.getLong("id")
            val type = jo.getString("type")
            val mBody = jo.getJSONObject("body")

            val hasPayload = jo.has("payloadSize")
            val payloadTransferInfo = if (hasPayload) jo.getJSONObject("payloadTransferInfo") else JSONObject()
            val payload = if (hasPayload) Payload(jo.getLong("payloadSize")) else null
            return NetworkPacket(id, type, mBody, payload, payloadTransferInfo)
        }
    }
}
